File: sdks/ios/Sources/PromptKeeper/SSEParser.swift
//
//  SSEParser.swift
//  PromptKeeper
//
//  Parses Server-Sent Events (data: ... lines, separated by blank line).
//

import Foundation

struct SSEDataEvent {
    var data: String?
}

enum SSEParser {
    /// Parse a single SSE message block (lines ending with blank line). Returns events (one per "data:" line).
    static func parse(_ block: String) -> [SSEDataEvent] {
        var events: [SSEDataEvent] = []
        var currentData: String?
        let lines = block.split(separator: "\n", omittingEmptySubsequences: false)
        for line in lines {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            if trimmed.isEmpty {
                if let d = currentData {
                    events.append(SSEDataEvent(data: d))
                    currentData = nil
                }
                continue
            }
            if trimmed.hasPrefix("data:") {
                let payload = String(trimmed.dropFirst(5)).trimmingCharacters(in: .whitespaces)
                currentData = payload.isEmpty ? nil : payload
            }
        }
        if let d = currentData {
            events.append(SSEDataEvent(data: d))
        }
        return events
    }

    /// If `data` is JSON of the form { "error": "..." }, returns the error message; otherwise nil.
    static func parseErrorPayload(_ data: String) -> String? {
        guard let json = data.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: json) as? [String: Any],
              let err = obj["error"] as? String else { return nil }
        return err
    }
}

File: sdks/android/src/main/kotlin/com/promptkeeper/sdk/sse/SSEParser.kt
package com.promptkeeper.sdk.sse

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.primitiveContent

/**
 * Parses Server-Sent Events (SSE) format: lines like "data: ...", separated by blank lines.
 */
internal object SSEParser {

    private val json = Json { ignoreUnknownKeys = true }

    /**
     * Parses a single SSE message block (ended by \n\n). Returns list of event data strings.
     */
    fun parse(block: String): List<String> {
        val result = mutableListOf<String>()
        var data = StringBuilder()
        for (line in block.split("\n")) {
            val trimmed = line.trim()
            when {
                trimmed.startsWith("data:") -> data.append(trimmed.removePrefix("data:").trim()).append("\n")
                trimmed.isEmpty() -> {
                    val d = data.toString().trim()
                    if (d.isNotEmpty()) result.add(d)
                    data = StringBuilder()
                }
            }
        }
        val d = data.toString().trim()
        if (d.isNotEmpty()) result.add(d)
        return result
    }

    /**
     * If [data] is JSON with an "error" field, returns that message; otherwise null.
     */
    fun parseErrorPayload(data: String): String? {
        return try {
            val obj = json.parseToJsonElement(data).jsonObject
            obj["error"]?.jsonPrimitive?.primitiveContent
        } catch (_: Exception) {
            null
        }
    }
}
